### PR TITLE
feature sync listen

### PR DESCRIPTION
**Pull request by ColinKotar**
Database sync using Github webhooks, this feature is for issue #14

### CODE REVIEW

File: gitkanban/sync.py [hunk 1]
```diff
@@ -1,29 +1,329 @@
+import json
+
 from sqlalchemy import create_engine
 from sqlalchemy.orm import sessionmaker
 
+import tornado.ioloop
+import tornado.web
+
+import requests
+from deeputil import Dummy
```
Comment (prashanthellina) [inline gitkanban/sync.py]: This had to be added as a dependency to setup.py

File: gitkanban/sync.py [hunk 2]
```diff
@@ -1,29 +1,329 @@
...
+        listen_cmd.set_defaults(func=self.cmd_listen)
+
+    def create_webhook(self, repo):
+        # creates a webhook for the specified repository
+
+        events = list(self.EVENT_FILTER_CONFIG.keys())
+
+        # FIXME: --webhook-url parameter should have this info
+        config = {
+            "url": "https://{host}.{endpoint}/listen".format(host=HOST, endpoint=ENDPOINT),
```
Comment (prashanthellina) [inline gitkanban/sync.py]: HOST and ENDPOINT should not be hard coded in this file. Take an argument called "--webhook-location". This should be a required parameter. eg value: http://example.com/. In our code, let us append '/listen' (using `urljoin` function from stdlib) before registering with Github.

Comment (ColinKotar) [reply gitkanban/sync.py]: Right I overlooked the HOST and ENDPOINT. Is this something I should take up now or are you making these changes?

**Comment by prashanthellina** [comment]
Looks like I’ll need time tomorrow to do the api anyway. You can make this
update then.

File: gitkanban/sync.py [hunk 3]
```diff
@@ -1,29 +1,329 @@
...
+            for label in event['issue']['labels']:
+                issue.labels.append(self.add_get_label(label))
+                self.session.commit()
+
+        elif action == 'unlabeled':
+            issue_node = event['issue']['node_id']
+            issue = self.session.query(Issue).filter_by(node_id=issue_node).first()
+            # change and commit all attributes
+            issue.labels.remove(self.add_get_label(event['label']))
+            self.session.commit()
```
Comment (prashanthellina) [inline gitkanban/sync.py]: @ColinKotar Please add a else clause here with an error log. An if elif construct must always have an else section.

File: gitkanban/sync.py (hunk 2 above)
Comment (prashanthellina) [reply gitkanban/sync.py]: I updated. A change to make it to make webhook loc an optional arg as a user might want to do a "full" sync without planning to run the "listen" service later.

File: gitkanban/sync.py (hunk 3 above)
Comment (ColinKotar) [reply gitkanban/sync.py]: @prashanthellina  I'm assuming I should add an error log at the end of each handle_x function? Such as, "error handling x event".

Comment (prashanthellina) [reply gitkanban/sync.py]: @ColinKotar yes. the else clause will be execute when none of the conditions in the if elif waterfall match. we can log something like self.log.error('unknown_action', event_name='....', action='....')

In general whenever you have if .. elif flow, we must have a closing else block.

File: gitkanban/sync.py [hunk 4]
```diff
@@ -1,29 +1,329 @@
...
+        org_node = event['organization']['node_id']
+
+        if action == 'deleted':
+            self.session.query(Organization).filter_by(node_id=org_node).delete()
+            self.session.commit()
+
+        if action == 'renamed':
+            org = self.session.query(Organization).filter_by(node_id=org_node).first()
+            org.name = event['organization']['name']
+            self.session.commit()
```
Comment (prashanthellina) [inline gitkanban/sync.py]: @ColinKotar Please convert thest two if blocks to a single if elif else construct.

File: gitkanban/sync.py [hunk 5]
```diff
@@ -299,16 +317,16 @@ def register(self, subcommands):
         subcommands = cmd.add_subparsers()
 
         full_cmd = subcommands.add_parser('full',
-                help='One-time full sync from Github via v3 API')
+                                          help='One-time full sync from Github via v3 API')
```
Comment (prashanthellina) [inline gitkanban/sync.py]: Can you move un-indent these like before? Is there a reason the help args are all indented differently or is the editor doing this automatically?

**Pull request merged**